### PR TITLE
Fix Fork Choice Reorg Conditions 

### PR DESCRIPTION
**Pull request by nisdas**
Only re-org if the new chain head isn't a direct child of the old head

### CODE REVIEW

File: beacon-chain/blockchain/fork_choice.go [hunk 1]
```diff
@@ -148,12 +149,21 @@ func (c *ChainService) ApplyForkChoiceRule(
...
+	if err != nil {
+		return fmt.Errorf("could not hash current chain head: %v", err)
+	}
 	c.canonicalBlocksLock.Lock()
 	defer c.canonicalBlocksLock.Unlock()
 	c.canonicalBlocks[head.Slot] = headRoot[:]
 
 	newState := postState
-	if head.Slot != block.Slot {
+	if chainRoot != bytesutil.ToBytes32(head.ParentRootHash32) {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/fork_choice.go]: A reorg happens when the new head does not meet these conditions
- newState != previousHead
- newState is some decedent of previousHead

File: beacon-chain/blockchain/fork_choice.go [hunk 2]
```diff
@@ -148,14 +149,28 @@ func (c *ChainService) ApplyForkChoiceRule(
 	if err != nil {
 		return fmt.Errorf("could not hash head block: %v", err)
 	}
+
+	chainHead, err := c.beaconDB.ChainHead()
```
Comment (terencechain) [inline beacon-chain/blockchain/fork_choice.go]:
```suggestion
	head, err := c.beaconDB.ChainHead()
```
prefer `head` or `blockHead` here

File: beacon-chain/blockchain/fork_choice.go [hunk 3]
```diff
@@ -148,14 +149,28 @@ func (c *ChainService) ApplyForkChoiceRule(
...
+	if err != nil {
+		return fmt.Errorf("could not retrieve chain head: %v", err)
+	}
+	chainRoot, err := hashutil.HashBeaconBlock(chainHead)
```
Comment (terencechain) [inline beacon-chain/blockchain/fork_choice.go]: `headRoot` sounds better here

File: beacon-chain/blockchain/fork_choice.go [hunk 4]
```diff
@@ -288,6 +312,24 @@ func (c *ChainService) blockChildren(ctx context.Context, block *pb.BeaconBlock,
 	return children, nil
 }
 
+func (c *ChainService) isDescendant(currentHead *pb.BeaconBlock, headRoot [32]byte,
```
Comment (terencechain) [inline beacon-chain/blockchain/fork_choice.go]: can you add a comment for `isDescendant`?

File: beacon-chain/blockchain/fork_choice.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/blockchain/fork_choice.go]: there would be two heads here then, one from lmdGhost and the other from `ChainHead()`

File: beacon-chain/blockchain/fork_choice.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/blockchain/fork_choice.go]: we already have a `headRoot` above, thats why I used chainHead to differentiate it

File: beacon-chain/blockchain/fork_choice.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/blockchain/fork_choice.go]: You are right, that was a bad suggestion

File: beacon-chain/blockchain/fork_choice_reorg_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"github.com/prysmaticlabs/prysm/shared/testutil"
+
```
Comment (terencechain) [inline beacon-chain/blockchain/fork_choice_reorg_test.go]:
```suggestion
```

File: beacon-chain/sync/receive_block.go [hunk 6]
```diff
@@ -205,6 +198,10 @@ func (rs *RegularSync) validateAndProcessBlock(
 		}
 	}
 
+	if err := rs.chainService.ApplyForkChoiceRule(ctx, block, beaconState); err != nil {
+		log.Errorf("Could not run fork choice on block %v", err)
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]:
```suggestion
		log.WithError(err).Error("Could not run fork choice on block")
```

**Pull request merged**